### PR TITLE
topology: tgl-max98373-rt5682: add smart amplifier support on sdw pla…

### PR DESCRIPTION
**Pull request by RanderWang**
…tforms

This patch is based on smart amplifier support on I2S platform (https://github.com/thesofproject/sof/pull/2829) and the
differences are pipeline layout and DAI config. On sdw platforms, the
pipeline layout is:

```
 PCM0 <---> volume <----> playback (Headset - ALC5682)
 PCM1 <---> volume <----> capture  (Headset - ALC5682)
 PCM2 ----> smart_amp ----> ALH0x102 (Speaker -max98373)
              ^
              |
              |
 PCM3 <---- demux  <----- ALH0x103 (Speaker -max98373)
 PCM4 <---- volume <---- DMIC01
 PCM5 <---- volume <---- DMIC16k
 PCM6 ----> volume -----> iDisp1
 PCM7 ----> volume -----> iDisp2
 PCM8 ----> volume -----> iDisp3
 PCM9 ----> volume -----> iDisp4
```

Signed-off-by: randerwang <rander.wang@linux.intel.com>

### CODE REVIEW

File: tools/topology/sof-tgl-sdw-max98373-rt5682.m4 [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+#
+# PCM0 <---> volume <----> playback (Headset - ALC5682)
+# PCM1 <---> volume <----> capture  (Headset - ALC5682)
+# PCM2 ----> smart_amp ----> ALH0x102 (Speaker -max98373)
+#              ^
+#              |
+#              |
+# PCM3 <---- demux  <----- ALH0x103 (Speaker -max98373)
+# PCM4 <---- volume <---- DMIC01
+# PCM5 <---- volume <---- DMIC16k
```
Comment (keyonjie) [inline tools/topology/sof-tgl-sdw-max98373-rt5682.m4:35]: Do we have WoV requirement for SDW platform also?

Comment (RanderWang) [reply tools/topology/sof-tgl-sdw-max98373-rt5682.m4:35]: Not now

Comment (keyonjie) [reply tools/topology/sof-tgl-sdw-max98373-rt5682.m4:35]: Okay, we can add it later.

File: tools/topology/sof-smart-amplifier.m4 [hunk 2]
```diff
@@ -131,19 +189,38 @@ SectionGraph."PIPE_SMART_AMP" {
...
 
 #
 # BE configurations - overrides config in ACPI if present
 #
 
+ifelse(SDW, `1',
+`
+#ALH ALH Pin2 (ID: SMART_BE_ID)
+DAI_CONFIG(ALH, SMART_ALH_INDEX, SMART_BE_ID, SMART_ALH_PLAYBACK_NAME,
+        ALH_CONFIG(ALH_CONFIG_DATA(ALH, SMART_ALH_INDEX, 48000, 2)))
```
Comment (keyonjie) [inline tools/topology/sof-smart-amplifier.m4:210]: ALH_CONFIG(ALH_CONFIG_DATA(ALH, SMART_ALH_INDEX, 48000, SMART_TX_CHANNELS))?

File: tools/topology/sof-smart-amplifier.m4 [hunk 3]
```diff
@@ -131,19 +189,38 @@ SectionGraph."PIPE_SMART_AMP" {
...
+
+#ALH ALH Pin3 (ID: SMART_BE_ID + 1)
+DAI_CONFIG(ALH, eval(SMART_ALH_INDEX + 1), eval(SMART_BE_ID + 1), SMART_ALH_CAPTURE_NAME,
+        ALH_CONFIG(ALH_CONFIG_DATA(ALH, eval(SMART_ALH_INDEX + 1), 48000, 2)))
```
Comment (keyonjie) [inline tools/topology/sof-smart-amplifier.m4:214]: ALH_CONFIG(ALH_CONFIG_DATA(ALH, SMART_ALH_INDEX, 48000, SMART_RX_CHANNELS))?

File: tools/topology/CMakeLists.txt [hunk 4]
```diff
@@ -130,6 +130,7 @@ set(TPLGS
 	"sof-imx8qxp-wm8960\;sof-imx8qxp-wm8960"
 	"sof-tgl-max98357a-rt5682\;sof-tgl-max98357a-rt5682"
 	"sof-tgl-max98373-rt5682\;sof-tgl-max98373-rt5682"
+	"sof-tgl-sdw-max98373-rt5682\;sof-tgl-sdw-max98373-rt5682\;-DSDW=1"
```
Comment (keyonjie) [inline tools/topology/CMakeLists.txt:133]: can we define SDW in the .m4 file as we don't need this to be configurable at compile time?

File: tools/topology/sof-smart-amplifier.m4 (hunk 3 above)
Comment (RanderWang) [reply tools/topology/sof-smart-amplifier.m4]: done

**Comment by RanderWang** [comment]
@keyonjie update my PR, thanks!

File: tools/topology/sof-tgl-sdw-max98373-rt5682.m4 [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+define(`SMART_PB_CH_NUM', 2)
+define(`SMART_TX_CHANNELS', 4)
+define(`SMART_RX_CHANNELS', 8)
+define(`SMART_FB_CHANNELS', 8)
+# Ref capture related
+define(`SMART_REF_PPL_ID', 4)
+define(`SMART_REF_CH_NUM', 4)
+# PCM related
+define(`SMART_PCM_ID', 2)
+define(`SMART_PCM_NAME', `smart373-spk')
```
Comment (keyonjie) [inline tools/topology/sof-tgl-sdw-max98373-rt5682.m4:64]: if you don't plan to use it, you can remove this definition?

**Comment by RanderWang** [comment]
@keyonjie   Thanks!  I update my PR.

**Comment by lgirdwood** [comment]
CI usual suspects.

**Pull request merged**